Implement shallow copy functions for `FunctionalTensorWrapper`.

ysiraichi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #118783

Fix: #115792

This PR implements 2 virtual functions of `TensorImpl` that are called when setting the
`tensor.data`:

- `shallow_copy_from`: which calls `copy_tensor_metadata`; and

- `copy_tensor_metadata`: which copies all `FunctionalTensorWrapper` metadata and ~calls
`dest->value_.set_data(src->value_)`~ assigns `dest->value_ = src->value_`, so as to copy also the inner tensor using the same
method

Before this PR, the inner tensor of a `FunctionalTensorWrapper` was being ignored.

cc @miladm @JackCaoG

ysiraichi committed:
Implement shallow copy functions for `FunctionalTensorWrapper`.

Fix: #115792

This PR implements 2 virtual functions of `TensorImpl` that are called when setting the
`tensor.data`:

- `shallow_copy_from`: which calls `copy_tensor_metadata`; and

- `copy_tensor_metadata`: which copies all `FunctionalTensorWrapper` metadata and calls
`dest->value_.set_data(src->value_)`, so as to copy also the inner tensor using the same
method

Before this PR, the inner tensor of a `FunctionalTensorWrapper` was being ignored.

[ghstack-poisoned]

ysiraichi commented on aten/src/ATen/FunctionalTensorWrapper.h:
```diff
@@ -230,6 +232,32 @@ struct TORCH_API FunctionalTensorWrapper : public c10::TensorImpl {
+
+    // FunctionalTensorWrapper-specific fields.
+    dest_impl->value_.set_data(src_impl->value_);
+    dest_impl->level_ = src_impl->level_;
+    dest_impl->mutation_counter_ = src_impl->mutation_counter_;
+    dest_impl->mutation_hidden_from_autograd_counter_ = src_impl->mutation_hidden_from_autograd_counter_;
+    dest_impl->mutation_during_no_grad_or_inference_mode_ = src_impl->mutation_during_no_grad_or_inference_mode_;
+    dest_impl->has_metadata_mutation_ = src_impl->has_metadata_mutation_;
+    dest_impl->is_multi_output_view_ = src_impl->is_multi_output_view_;
+    dest_impl->was_storage_changed_ = src_impl->was_storage_changed_;
+    dest_impl->generation_ = src_impl->generation_;
+    dest_impl->view_metas_ = src_impl->view_metas_;
```
@bdhirsh Do you think it's ok to copy all these metadata? Just to be sure I'm not doing anything stupid, here.

ysiraichi committed:
Update on "Implement shallow copy functions for `FunctionalTensorWrapper`."


Fix: #115792

This PR implements 2 virtual functions of `TensorImpl` that are called when setting the
`tensor.data`:

- `shallow_copy_from`: which calls `copy_tensor_metadata`; and

- `copy_tensor_metadata`: which copies all `FunctionalTensorWrapper` metadata and calls
`dest->value_.set_data(src->value_)`, so as to copy also the inner tensor using the same
method

Before this PR, the inner tensor of a `FunctionalTensorWrapper` was being ignored.

cc miladm JackCaoG 

[ghstack-poisoned]

ysiraichi committed:
Fix lint and failing tests. on "Implement shallow copy functions for `FunctionalTensorWrapper`."


Fix: #115792

This PR implements 2 virtual functions of `TensorImpl` that are called when setting the
`tensor.data`:

- `shallow_copy_from`: which calls `copy_tensor_metadata`; and

- `copy_tensor_metadata`: which copies all `FunctionalTensorWrapper` metadata and calls
`dest->value_.set_data(src->value_)`, so as to copy also the inner tensor using the same
method

Before this PR, the inner tensor of a `FunctionalTensorWrapper` was being ignored.

cc miladm JackCaoG 

[ghstack-poisoned]

ysiraichi commented:
Replaced `dest->value_.set_data(src->value_)` for plain assignment `dest->value_ = src->value_` due to 2 reasons:

- `test_functionalize_grad_cpu` was failing due to metadata change during functionalization
- `FunctionalTensorWrapper::shallow_copy_and_detach_core` used the same `value_` tensor for creating a new one

lezcano commented:
Adding @albanD and @mikaylagawarecki as reviewers, as this is a more targeted fix for https://github.com/pytorch/pytorch/issues/115792. I wonder whether we still want https://github.com/pytorch/pytorch/pull/117167 after this fix.

ysiraichi commented:
@albanD @bdhirsh This is a friendly reminder. Do you have some time to take a look at this PR?

bdhirsh reviewed (approved):
Sorry for the delay. I think this looks reasonable to unblock `module.to(device)` for XLA!

mikaylagawarecki commented:
@pytorchbot merge -r

ysiraichi committed:
Update on "Implement shallow copy functions for `FunctionalTensorWrapper`."


Fix: #115792

This PR implements 2 virtual functions of `TensorImpl` that are called when setting the
`tensor.data`:

- `shallow_copy_from`: which calls `copy_tensor_metadata`; and

- `copy_tensor_metadata`: which copies all `FunctionalTensorWrapper` metadata and ~calls
`dest->value_.set_data(src->value_)`~ assigns `dest->value_ = src->value_`, so as to copy also the inner tensor using the same
method

Before this PR, the inner tensor of a `FunctionalTensorWrapper` was being ignored.

cc miladm JackCaoG 

[ghstack-poisoned]

mikaylagawarecki commented:
@ysiraichi Hm the windows failure on trunk seems related :/

ysiraichi committed:
Fix regex. on "Implement shallow copy functions for `FunctionalTensorWrapper`."


Fix: #115792

This PR implements 2 virtual functions of `TensorImpl` that are called when setting the
`tensor.data`:

- `shallow_copy_from`: which calls `copy_tensor_metadata`; and

- `copy_tensor_metadata`: which copies all `FunctionalTensorWrapper` metadata and ~calls
`dest->value_.set_data(src->value_)`~ assigns `dest->value_ = src->value_`, so as to copy also the inner tensor using the same
method

Before this PR, the inner tensor of a `FunctionalTensorWrapper` was being ignored.

cc miladm JackCaoG 

[ghstack-poisoned]

ysiraichi commented:
@pytorchbot merge
